src/IECoreDelight/Renderer.cpp

```diff
@@ -389,15 +396,185 @@ IE_CORE_DECLAREPTR( DelightOutput )
 namespace
 {
 
-class DelightShader : public IECore::RefCounted
+// From https://gitlab.com/3Delight/3delight-for-houdini/-/blob/master/osl_utilities.cpp
+enum BasisTypes
+{
+	CONSTANT,
+	LINEAR,
+	MONOTONECUBIC,
+	CATMULLROM
+};
+
+int basisInt( const std::string &basis )
+{
+	if( basis == "constant" )
+	{
+		return BasisTypes::CONSTANT;
+	}
+	if( basis == "linear" )
+	{
+		return BasisTypes::LINEAR;
+	}
+	// `SplinePlug` converts from `monotonecubic` to `bezier`, so we'll never get `monotonecubic`
+
+	return BasisTypes::CATMULLROM;
+}
+
+const OSL::OSLQuery::Parameter *splineValueParameter(
+	const OSL::OSLQuery &query,
+	const std::string &splineParameterName
+)
+{
+	for( size_t i = 0, eI = query.nparams(); i < eI; ++i )
+	{
+		const OSL::OSLQuery::Parameter *p = query.getparam( i );
+
+		if( !boost::starts_with( p->name, splineParameterName ) )
+		{
+			continue;
+		}
+		for( const auto &m : p->metadata )
+		{
+			if(
+				m.name == "widget" &&
+				m.sdefault.size() > 0 &&
+				m.sdefault[0].find( "Ramp" ) != std::string::npos
+			)
+			{
+				return p;
+			}
+		}
+	}
+	return nullptr;
+}
+
+bool find3DelightSplineParameters(
+	const OSL::OSLQuery &query,
+	const std::string &splineParameterName,
+	const OSL::OSLQuery::Parameter * &positionsParameter,
+	const OSL::OSLQuery::Parameter * &valuesParameter,
+	const OSL::OSLQuery::Parameter * &basisParameter
+)
+{
+	positionsParameter = nullptr;
+	valuesParameter = nullptr;
+	basisParameter = nullptr;
+
+	valuesParameter = splineValueParameter( query, splineParameterName );
+
+	if( valuesParameter )
+	{
+		for( size_t i = 0, eI = query.nparams(); i < eI; ++i )
+		{
+			const OSL::OSLQuery::Parameter *p = query.getparam( i );
+			if( p == valuesParameter || !p->type.is_array() || !boost::starts_with( p->name, splineParameterName ) )
+			{
+				continue;
+			}
+			if( p->type.basetype == OSL::TypeDesc::INT && p->type.aggregate == OSL::TypeDesc::SCALAR )
+			{
+				// Here we prefer the `int` value basis parameter because that is the only
+				// basis parameter that is consistently found in all 3delight splines.
+				basisParameter = p;
+			}
+			if( p->type.basetype == OSL::TypeDesc::FLOAT && p->type.aggregate == OSL::TypeDesc::SCALAR )
+			{
+				positionsParameter = p;
+			}
+		}
+	}
+
+	return positionsParameter && valuesParameter && basisParameter;
+}
+
+ConstCompoundDataPtr renameParameters( const std::string &shaderFileName, const ConstCompoundDataPtr &parametersData )
```

Comment:
I haven't reviewed this closely yet, or tested the results on 3Delight's splines, but Gaffer splines do seem to be broken. In the example below, I always get a default black-to-white ramp, no matter how many knots I add.

<details>

```
import Gaffer
import GafferDelight
import GafferImage
import GafferOSL
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 3, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["InteractiveDelightRender"] = GafferDelight.InteractiveDelightRender( "InteractiveDelightRender" )
parent.addChild( __children["InteractiveDelightRender"] )
__children["InteractiveDelightRender"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"] = GafferScene.Outputs( "Outputs" )
parent.addChild( __children["Outputs"] )
__children["Outputs"]["outputs"].addChild( Gaffer.ValuePlug( "output1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "name", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.BoolPlug( "active", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "fileName", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "type", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.StringPlug( "data", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"].addChild( Gaffer.CompoundDataPlug( "parameters", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "driverType", Gaffer.StringPlug( "value", defaultValue = 'ClientDisplayDriver', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "driverType", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "displayHost", Gaffer.StringPlug( "value", defaultValue = 'localhost', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "displayHost", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "displayPort", Gaffer.StringPlug( "value", defaultValue = '${image:catalogue:port}', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "displayPort", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "remoteDisplayType", Gaffer.StringPlug( "value", defaultValue = 'GafferImage::GafferDisplayDriver', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "remoteDisplayType", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "catalogue:imageName", Gaffer.StringPlug( "value", defaultValue = 'Image', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "catalogue_imageName", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"]["outputs"]["output1"]["parameters"].addChild( Gaffer.NameValuePlug( "quantize", Gaffer.IntVectorDataPlug( "value", defaultValue = IECore.IntVectorData( [ 0, 0, 0, 0 ] ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "quantize", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Camera"] = GafferScene.Camera( "Camera" )
parent.addChild( __children["Camera"] )
__children["Camera"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["StandardOptions"] = GafferScene.StandardOptions( "StandardOptions" )
parent.addChild( __children["StandardOptions"] )
__children["StandardOptions"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"] = GafferImage.Catalogue( "Catalogue" )
parent.addChild( __children["Catalogue"] )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"]["images"].addChild( GafferImage.Catalogue.Image( "Image3", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Catalogue"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Plane"] = GafferScene.Plane( "Plane" )
parent.addChild( __children["Plane"] )
__children["Plane"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderAssignment1"] = GafferScene.ShaderAssignment( "ShaderAssignment1" )
parent.addChild( __children["ShaderAssignment1"] )
__children["ShaderAssignment1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Constant"] = GafferOSL.OSLShader( "Constant" )
parent.addChild( __children["Constant"] )
__children["Constant"].loadShader( "Surface/Constant" )
__children["Constant"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"] = GafferOSL.OSLShader( "ColorSpline" )
parent.addChild( __children["ColorSpline"] )
__children["ColorSpline"].loadShader( "Pattern/ColorSpline" )
__children["ColorSpline"]["parameters"]["spline"].clearPoints()
__children["ColorSpline"]["parameters"]["spline"].addChild( Gaffer.ValuePlug( "p0", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p0"].addChild( Gaffer.FloatPlug( "x", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p0"].addChild( Gaffer.Color3fPlug( "y", defaultValue = imath.Color3f( 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"].addChild( Gaffer.ValuePlug( "p1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p1"].addChild( Gaffer.FloatPlug( "x", defaultValue = 1.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p1"].addChild( Gaffer.Color3fPlug( "y", defaultValue = imath.Color3f( 1, 1, 1 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"].addChild( Gaffer.ValuePlug( "p4", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p4"].addChild( Gaffer.FloatPlug( "x", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p4"].addChild( Gaffer.Color3fPlug( "y", defaultValue = imath.Color3f( 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"].addChild( Gaffer.ValuePlug( "p5", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p5"].addChild( Gaffer.FloatPlug( "x", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"]["parameters"]["spline"]["p5"].addChild( Gaffer.Color3fPlug( "y", defaultValue = imath.Color3f( 0, 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ColorSpline"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["InteractiveDelightRender"]["in"].setInput( __children["StandardOptions"]["out"] )
__children["InteractiveDelightRender"]["__uiPosition"].setValue( imath.V2f( 26.9985447, -16.2680244 ) )
__children["Outputs"]["in"].setInput( __children["Group"]["out"] )
__children["Outputs"]["outputs"]["output1"]["name"].setValue( 'Interactive/Beauty' )
__children["Outputs"]["outputs"]["output1"]["fileName"].setValue( 'beauty' )
__children["Outputs"]["outputs"]["output1"]["type"].setValue( 'ieDisplay' )
__children["Outputs"]["outputs"]["output1"]["data"].setValue( 'rgba' )
__children["Outputs"]["__uiPosition"].setValue( imath.V2f( 27.0039635, 0.0601475127 ) )
__children["Camera"]["transform"]["translate"].setValue( imath.V3f( 0, 0, 2.11595345 ) )
__children["Camera"]["__uiPosition"].setValue( imath.V2f( 18.2114639, 16.3883095 ) )
__children["Group"]["in"][0].setInput( __children["Camera"]["out"] )
__children["Group"]["in"][1].setInput( __children["ShaderAssignment1"]["out"] )
__children["Group"]["__uiPosition"].setValue( imath.V2f( 27.008112, 8.22424698 ) )
__children["StandardOptions"]["in"].setInput( __children["Outputs"]["out"] )
__children["StandardOptions"]["options"]["renderCamera"]["value"].setValue( '/group/camera' )
__children["StandardOptions"]["options"]["renderCamera"]["enabled"].setValue( True )
__children["StandardOptions"]["__uiPosition"].setValue( imath.V2f( 27.0005035, -8.10394382 ) )
__children["Catalogue"]["images"]["Image"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/delightVolumes/5530c425693b7709f90a9f0e18b78b9b.exr' )
__children["Catalogue"]["images"]["Image1"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/delightVolumes/a8b418f564cde3a3b9a0e32eaa22cb8a.exr' )
__children["Catalogue"]["images"]["Image2"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/delightVolumes/13f0fc740dc4cd3f7bfc897bb7f252cf.exr' )
__children["Catalogue"]["images"]["Image3"]["fileName"].setValue( '/home/john/gaffer/projects/default/catalogues/delightVolumes/13f0fc740dc4cd3f7bfc897bb7f252cf.exr' )
__children["Catalogue"]["imageIndex"].setValue( 4 )
__children["Catalogue"]["directory"].setValue( '${project:rootDirectory}/catalogues/${script:name}' )
__children["Catalogue"]["__uiPosition"].setValue( imath.V2f( 46.0056915, -15.9897099 ) )
__children["Plane"]["__uiPosition"].setValue( imath.V2f( 33.9223442, 34.4282455 ) )
__children["ShaderAssignment1"]["in"].setInput( __children["Plane"]["out"] )
__children["ShaderAssignment1"]["shader"].setInput( __children["Constant"]["out"] )
__children["ShaderAssignment1"]["__uiPosition"].setValue( imath.V2f( 33.9213867, 26.2641582 ) )
__children["Constant"]["parameters"]["Cs"].setInput( __children["ColorSpline"]["out"]["c"] )
__children["Constant"]["__uiPosition"].setValue( imath.V2f( 7.82520437, 26.2641964 ) )
__children["ColorSpline"]["parameters"]["spline"]["interpolation"].setValue( 2 )
__children["ColorSpline"]["parameters"]["spline"]["p4"]["x"].setValue( 0.23097112774848938 )
__children["ColorSpline"]["parameters"]["spline"]["p4"]["y"].setValue( imath.Color3f( 0.0474949442, 0.0648148134, 0.23884511 ) )
__children["ColorSpline"]["parameters"]["spline"]["p5"]["x"].setValue( 0.5125628113746643 )
__children["ColorSpline"]["parameters"]["spline"]["p5"]["y"].setValue( imath.Color3f( 0.135015815, 0.146225914, 0.258865207 ) )
__children["ColorSpline"]["parameters"]["direction"].setValue( 'u' )
__children["ColorSpline"]["__uiPosition"].setValue( imath.V2f( -4.8525219, 26.2642155 ) )


del __children


```
</details>

Reply:
Fixed in the latest push